pkg/auth/providers/migration/activedirectory_migration.go

```diff
@@ -228,3 +232,19 @@ func (m *adMigration) migratePRTB(newPrincipalID string, dn string) error {
 	}
 	return nil
 }
+
+func getPrincipalWithRetry(adProvider common.AuthProvider, dn string, token v3.Token) (*v3.Principal, error) {
+	for retry := 0; retry < maxRetries; retry++ {
+		userPrincipal, err := adProvider.GetPrincipal(dn, token)
+		if err == nil {
+			return &userPrincipal, nil
+		}
+		var apiError *httperror.APIError
+		if errors.As(err, &apiError) && httperror.IsNotFound(apiError) {
+			return nil, err
+		}
+		logrus.Debugf("fetch principal failed: %v. Retrying in %v...\n", err, retryDelay)
+		time.Sleep(retryDelay)
+	}
+	return nil, fmt.Errorf("failed to fetch principal after %d retries", maxRetries)
+}
```

Comment:
@crobby Can you elaborate on the reasons why getPrincipal can fails and why a retry might help?

Reply:
Under the hood, it is making a call to LDAP.  In my testing, I've seen that ldap call return an error.  In all cases where I've seen it happen, a few retries spaced out a bit has resolved the issue.